File: src/uimd/runtime/style.py
from uimd.runtime.uiinstance import UIInstance


DEFAULT_BORDER_WIDTH_HORIZONTAL = 0
DEFAULT_BORDER_WIDTH_VERTICAL = 0
DEFAULT_SCOPE_DIM_BACKGROUND = "#00000066"


class Color:
    """Represents a color with r, g, b values (0-255) or a hex string."""

    def __init__(self, value=None, alpha=1.0):
        if isinstance(value, str):
            value = value.strip()
            if value.endswith(";"):
                value = value[:-1].rstrip()
            if value == "transparent":
                self.hex = "transparent"
                self.r, self.g, self.b = -1, -1, -1
                self.alpha = 0.0
            elif len(value) == 4:
                expanded = "#" + value[1]*2 + value[2]*2 + value[3]*2
                self.hex = expanded
                r = int(expanded[1:3], 16)
                g = int(expanded[3:5], 16)
                b = int(expanded[5:7], 16)
                self.r, self.g, self.b = r, g, b
            elif len(value) == 5:
                expanded = "#" + value[1]*2 + value[2]*2 + value[3]*2
                self.hex = expanded + value[4]*2
                r = int(expanded[1:3], 16)
                g = int(expanded[3:5], 16)
                b = int(expanded[5:7], 16)
                a = int(value[4]*2, 16) / 255
                self.r, self.g, self.b = r, g, b
                self.alpha = a
            elif len(value) == 9:
                self.hex = value
                r = int(value[1:3], 16)
                g = int(value[3:5], 16)
                b = int(value[5:7], 16)
                a = int(value[7:9], 16) / 255
                self.r, self.g, self.b = r, g, b
                self.alpha = a
            else:
                self.hex = value
                r = int(value[1:3], 16)
                g = int(value[3:5], 16)
                b = int(value[5:7], 16)
                self.r, self.g, self.b = r, g, b
        elif isinstance(value, Color):
            self.hex = value.hex
            self.r, self.g, self.b = value.r, value.g, value.b
            self.alpha = value.alpha
        else:
            self.hex = "#ffffff"
            self.r, self.g, self.b = 255, 255, 255
        if not hasattr(self, 'alpha'):
            self.alpha = alpha

    def __eq__(self, other):
        if isinstance(other, Color):
            return self.hex == other.hex and abs(self.alpha - other.alpha) < 0.001
        return False

    def __repr__(self):
        if self.alpha < 1.0:
            return f"Color({self.hex}, alpha={self.alpha:.2f})"
        return f"Color({self.hex})"


class Style:
    """Style definition for a UI element or window."""

    @staticmethod
    def _lighten(hex_color, percent):
        r = int(hex_color[1:3], 16)
        g = int(hex_color[3:5], 16)
        b = int(hex_color[5:7], 16)
        r = min(255, round(r + (255 - r) * percent))
        g = min(255, round(g + (255 - g) * percent))
        b = min(255, round(b + (255 - b) * percent))
        return f"#{r:02x}{g:02x}{b:02x}"

    @staticmethod
    def _blend_color_over(color, background):
        if color is None:
            return background
        alpha = getattr(color, "alpha", 1.0)
        if alpha is None or alpha >= 1.0:
            return color
        if alpha <= 0.0:
            return background
        if background is None or getattr(background, "r", -1) < 0:
            return color
        r = round(alpha * color.r + (1 - alpha) * background.r)
        g = round(alpha * color.g + (1 - alpha) * background.g)
        b = round(alpha * color.b + (1 - alpha) * background.b)
        return Color(f"#{r:02x}{g:02x}{b:02x}")

    def __init__(self):
        self.color = None  # Color for text
        self.background = None  # Color for background
        self.background_texture = ""
        self.background_texture_color = None
        self.user_select = None
        self.scroll_x = False
        self.scroll_y = True
        self.color_alpha = None  # Alpha for text color (0.0-1.0)
        self.background_alpha = None  # Alpha for background color (0.0-1.0)
        self.padding = (0, 0, 0, 0)  # top, right, bottom, left
        self.margin = (0, 0, 0, 0)  # top, right, bottom, left
        self.border_color = None
        self.border_width_horizontal = DEFAULT_BORDER_WIDTH_HORIZONTAL
        self.border_width_vertical = DEFAULT_BORDER_WIDTH_VERTICAL
        self.focus_background = None
        self.focus_color = None
        self.text_align = None
        self.gap = 0
        self.word_wrap = True
        self.text_color_gradient = None
        self.text_background_gradient = None
        self.scope_dim_background = Color(DEFAULT_SCOPE_DIM_BACKGROUND)

    @property
    def border_width(self):
        return self.border_width_horizontal

    @border_width.setter
    def border_width(self, value):
        width = int(value)
        self.border_width_horizontal = width
        self.border_width_vertical = width

    @staticmethod
    def _parse_box_values(value):
        parts = str(value).split(",")
        nums = tuple(int(p.strip()) for p in parts)
        if len(nums) == 1:
            return (nums[0], nums[0], nums[0], nums[0])
        if len(nums) == 2:
            return (nums[0], nums[1], nums[0], nums[1])
        if len(nums) == 4:
            return nums
        return nums + (0,) * (4 - len(nums))

    @staticmethod
    def _set_box_side(values, side, value):
        top, right, bottom, left = values
        parsed = int(value)
        if side == "top":
            top = parsed
        elif side == "right":
            right = parsed
        elif side == "bottom":
            bottom = parsed
        elif side == "left":
            left = parsed
        return (top, right, bottom, left)

    def set(self, key, value):
        if key == "color":
            self.color = Color(value)
        elif key == "background":
            self.background = Color(value)
        elif key == "background-texture":
            self.background_texture = str(value)
        elif key == "background-texture-color":
            self.background_texture_color = Color(value)
        elif key == "user-select":
            self.user_select = str(value).strip().lower()
        elif key == "scroll-x":
            if isinstance(value, bool):
                self.scroll_x = value
            else:
                self.scroll_x = str(value).strip().lower() in ("true", "yes", "1")
        elif key == "scroll-y":
            if isinstance(value, bool):
                self.scroll_y = value
            else:
                self.scroll_y = str(value).strip().lower() in ("true", "yes", "1")
        elif key == "padding":
            self.padding = self._parse_box_values(value)
        elif key == "margin":
            self.margin = self._parse_box_values(value)
        elif key in ("margin-top", "top-margin"):
            self.margin = self._set_box_side(self.margin, "top", value)
        elif key in ("margin-right", "right-margin"):
            self.margin = self._set_box_side(self.margin, "right", value)
        elif key in ("margin-bottom", "bottom-margin"):
            self.margin = self._set_box_side(self.margin, "bottom", value)
        elif key in ("margin-left", "left-margin"):
            self.margin = self._set_box_side(self.margin, "left", value)
        elif key == "color-alpha":
            self.color_alpha = float(value)
        elif key == "background-alpha":
            self.background_alpha = float(value)
        elif key == "border-color":
            self.border_color = Color(value)
        elif key == "border-width":
            self.border_width = value
        elif key == "border-width-horizontal":
            self.border_width_horizontal = int(value)
        elif key == "border-width-vertical":
            self.border_width_vertical = int(value)
        elif key == "focus-background":
            self.focus_background = Color(value)
        elif key == "focus-color":
            self.focus_color = Color(value)
        elif key == "text-align":
            self.text_align = str(value).strip().lower()
        elif key == "gap":
            self.gap = int(value)
        elif key == "word-wrap":
            if isinstance(value, bool):
                self.word_wrap = value
            elif isinstance(value, str):
                self.word_wrap = value.lower() in ("true", "yes", "1")
        elif key == "text-color-gradient":
            self.text_color_gradient = self._parse_text_gradient(value)
        elif key == "text-background-gradient":
            self.text_background_gradient = self._parse_text_gradient(value)
        elif key == "scope-dim-background":
            self.scope_dim_background = Color(value)

    @staticmethod
    def _parse_text_gradient(value):
        if not isinstance(value, dict):
            return None
        colors = value.get("colors", [])
        if isinstance(colors, str):
            colors = [colors]
        parsed_colors = [Color(color) for color in colors if str(color).strip()]
        if not parsed_colors:
            return None
        return {
            "interval": max(1, int(value.get("interval", 70) or 70)),
            "step": int(value.get("step", 1) or 1),
            "segment_size": max(1, int(value.get("segment-size", value.get("segment_size", 1)) or 1)),
            "colors": parsed_colors,
        }

    def copy(self):
        s = Style()
        s.color = self.color
        s.background = self.background
        s.background_texture = self.background_texture
        s.background_texture_color = self.background_texture_color
        s.user_select = self.user_select
        s.scroll_x = self.scroll_x
        s.scroll_y = self.scroll_y
        s.color_alpha = self.color_alpha
        s.background_alpha = self.background_alpha
        s.padding = self.padding
        s.margin = self.margin
        s.border_color = self.border_color
        s.border_width_horizontal = self.border_width_horizontal
        s.border_width_vertical = self.border_width_vertical
        s.focus_background = self.focus_background
        s.focus_color = self.focus_color
        s.text_align = self.text_align
        s.gap = self.gap
        s.word_wrap = self.word_wrap
        s.text_color_gradient = self.text_color_gradient
        s.text_background_gradient = self.text_background_gradient
        s.scope_dim_background = self.scope_dim_background
        return s

    def merge(self, other):
        """Merge another style on top of this one (non-None values override)."""
        if other.color is not None:
            self.color = other.color
        if other.background is not None:
            self.background = other.background
        if other.background_texture:
            self.background_texture = other.background_texture
        if other.background_texture_color is not None:
            self.background_texture_color = other.background_texture_color
        if other.user_select is not None:
            self.user_select = other.user_select
        if other.scroll_x:
            self.scroll_x = other.scroll_x
        if not other.scroll_y:
            self.scroll_y = other.scroll_y
        if other.color_alpha is not None:
            self.color_alpha = other.color_alpha
        if other.background_alpha is not None:
            self.background_alpha = other.background_alpha
        if other.padding != (0, 0, 0, 0):
            self.padding = other.padding
        if other.margin != (0, 0, 0, 0):
            self.margin = other.margin
        if other.border_color is not None:
            self.border_color = other.border_color
        if other.border_width_horizontal != DEFAULT_BORDER_WIDTH_HORIZONTAL:
            self.border_width_horizontal = other.border_width_horizontal
        if other.border_width_vertical != DEFAULT_BORDER_WIDTH_VERTICAL:
            self.border_width_vertical = other.border_width_vertical
        if other.focus_background is not None:
            self.focus_background = other.focus_background
        if other.focus_color is not None:
            self.focus_color = other.focus_color
        if other.text_align is not None:
            self.text_align = other.text_align
        if other.gap != 0:
            self.gap = other.gap
        if other.word_wrap is not None:
            self.word_wrap = other.word_wrap
        if other.text_color_gradient is not None:
            self.text_color_gradient = other.text_color_gradient
        if other.text_background_gradient is not None:
            self.text_background_gradient = other.text_background_gradient
        if other.scope_dim_background is not None:
            self.scope_dim_background = other.scope_dim_background


DEFAULT_TEXT_COLOR = Color("#ffffff")
DEFAULT_BG_COLOR = Color("#000000")
DEFAULT_FOCUS_BG_COLOR = Color("#444444")

File: java/src/main/java/uimd/Style.java
package uimd;

public final class Style
{
    private Color color;
    private Color background;
    private String backgroundTexture;
    private Color backgroundTextureColor;
    private Color scopeDimBackground = new Color("#00000066");
    private Color borderColor;
    private Integer borderWidthHorizontal;
    private Integer borderWidthVertical;
    private Integer padding;
    private Integer paddingTop;
    private Integer paddingRight;
    private Integer paddingBottom;
    private Integer paddingLeft;
    private Integer margin;
    private Integer marginTop;
    private Integer marginRight;
    private Integer marginBottom;
    private Integer marginLeft;
    private Integer gap;
    private String textAlign = "";
    private String userSelect = "";
    private Boolean scrollX;
    private Boolean scrollY;
    private TextGradient textColorGradient;
    private TextGradient textBackgroundGradient;

    public static Style fromProperties(String... properties)
    {
        if (properties == null || properties.length % 2 != 0)
        {
            throw new IllegalArgumentException("style properties must be key/value pairs");
        }
        Style result = new Style();
        for (int index = 0; index < properties.length; index += 2)
        {
            String key = properties[index];
            String value = properties[index + 1];
            switch (key)
            {
                case "color" -> result.setColor(new Color(value));
                case "background" -> result.setBackground(new Color(value));
                case "background-texture" -> result.setBackgroundTexture(value);
                case "background-texture-color" -> result.setBackgroundTextureColor(new Color(value));
                case "scope-dim-background" -> result.setScopeDimBackground(new Color(value));
                case "border-color" -> result.setBorderColor(new Color(value));
                case "border-width-horizontal" -> result.setBorderWidthHorizontal(parseInteger(value));
                case "border-width-vertical" -> result.setBorderWidthVertical(parseInteger(value));
                case "padding" -> result.setPadding(parseInteger(value));
                case "padding-top" -> result.setPaddingTop(parseInteger(value));
                case "padding-right" -> result.setPaddingRight(parseInteger(value));
                case "padding-bottom" -> result.setPaddingBottom(parseInteger(value));
                case "padding-left" -> result.setPaddingLeft(parseInteger(value));
                case "margin" -> result.setMargin(parseInteger(value));
                case "margin-top" -> result.setMarginTop(parseInteger(value));
                case "margin-right" -> result.setMarginRight(parseInteger(value));
                case "margin-bottom" -> result.setMarginBottom(parseInteger(value));
                case "margin-left" -> result.setMarginLeft(parseInteger(value));
                case "gap" -> result.setGap(parseInteger(value));
                case "text-align" -> result.setTextAlign(value);
                case "user-select" -> result.setUserSelect(value);
                case "scroll-x" -> result.setScrollX(parseBoolean(value));
                case "scroll-y" -> result.setScrollY(parseBoolean(value));
                default -> throw new IllegalArgumentException("unsupported style property: " + key);
            }
        }
        return result;
    }

    public Style copy()
    {
        Style result = new Style();
        result.color = color;
        result.background = background;
        result.backgroundTexture = backgroundTexture;
        result.backgroundTextureColor = backgroundTextureColor;
        result.scopeDimBackground = scopeDimBackground;
        result.borderColor = borderColor;
        result.borderWidthHorizontal = borderWidthHorizontal;
        result.borderWidthVertical = borderWidthVertical;
        result.padding = padding;
        result.paddingTop = paddingTop;
        result.paddingRight = paddingRight;
        result.paddingBottom = paddingBottom;
        result.paddingLeft = paddingLeft;
        result.margin = margin;
        result.marginTop = marginTop;
        result.marginRight = marginRight;
        result.marginBottom = marginBottom;
        result.marginLeft = marginLeft;
        result.gap = gap;
        result.textAlign = textAlign;
        result.userSelect = userSelect;
        result.scrollX = scrollX;
        result.scrollY = scrollY;
        result.textColorGradient = textColorGradient;
        result.textBackgroundGradient = textBackgroundGradient;
        return result;
    }

    public void merge(Style other)
    {
        if (other == null)
        {
            return;
        }
        color = other.color == null ? color : other.color;
        background = other.background == null ? background : other.background;
        backgroundTexture = other.backgroundTexture == null ? backgroundTexture : other.backgroundTexture;
        backgroundTextureColor = other.backgroundTextureColor == null ? backgroundTextureColor : other.backgroundTextureColor;
        scopeDimBackground = other.scopeDimBackground == null ? scopeDimBackground : other.scopeDimBackground;
        borderColor = other.borderColor == null ? borderColor : other.borderColor;
        borderWidthHorizontal = other.borderWidthHorizontal == null ? borderWidthHorizontal : other.borderWidthHorizontal;
        borderWidthVertical = other.borderWidthVertical == null ? borderWidthVertical : other.borderWidthVertical;
        padding = other.padding == null ? padding : other.padding;
        paddingTop = other.paddingTop == null ? paddingTop : other.paddingTop;
        paddingRight = other.paddingRight == null ? paddingRight : other.paddingRight;
        paddingBottom = other.paddingBottom == null ? paddingBottom : other.paddingBottom;
        paddingLeft = other.paddingLeft == null ? paddingLeft : other.paddingLeft;
        margin = other.margin == null ? margin : other.margin;
        marginTop = other.marginTop == null ? marginTop : other.marginTop;
        marginRight = other.marginRight == null ? marginRight : other.marginRight;
        marginBottom = other.marginBottom == null ? marginBottom : other.marginBottom;
        marginLeft = other.marginLeft == null ? marginLeft : other.marginLeft;
        gap = other.gap == null ? gap : other.gap;
        textAlign = other.textAlign == null || other.textAlign.isEmpty() ? textAlign : other.textAlign;
        userSelect = other.userSelect == null || other.userSelect.isEmpty() ? userSelect : other.userSelect;
        scrollX = other.scrollX == null ? scrollX : other.scrollX;
        scrollY = other.scrollY == null ? scrollY : other.scrollY;
        textColorGradient = other.textColorGradient == null ? textColorGradient : other.textColorGradient;
        textBackgroundGradient = other.textBackgroundGradient == null ? textBackgroundGradient : other.textBackgroundGradient;
    }

    public Style merged(Style other)
    {
        Style result = copy();
        result.merge(other);
        return result;
    }

    public Color color()
    {
        return color;
    }

    public void setColor(Color value)
    {
        color = value;
    }

    public Color background()
    {
        return background;
    }

    public void setBackground(Color value)
    {
        background = value;
    }

    public String backgroundTexture()
    {
        return backgroundTexture;
    }

    public void setBackgroundTexture(String value)
    {
        backgroundTexture = value;
    }

    public Color backgroundTextureColor()
    {
        return backgroundTextureColor;
    }

    public void setBackgroundTextureColor(Color value)
    {
        backgroundTextureColor = value;
    }

    public Color scopeDimBackground()
    {
        return scopeDimBackground;
    }

    public void setScopeDimBackground(Color value)
    {
        scopeDimBackground = value;
    }

    public Color borderColor()
    {
        return borderColor;
    }

    public void setBorderColor(Color value)
    {
        borderColor = value;
    }

    public Integer borderWidthHorizontal()
    {
        return borderWidthHorizontal;
    }

    public void setBorderWidthHorizontal(Integer value)
    {
        borderWidthHorizontal = value;
    }

    public Integer borderWidthVertical()
    {
        return borderWidthVertical;
    }

    public void setBorderWidthVertical(Integer value)
    {
        borderWidthVertical = value;
    }

    public Integer padding()
    {
        return padding;
    }

    public void setPadding(Integer value)
    {
        padding = value;
    }

    public Integer paddingTop()
    {
        return paddingTop;
    }

    public void setPaddingTop(Integer value)
    {
        paddingTop = value;
    }

    public Integer paddingRight()
    {
        return paddingRight;
    }

    public void setPaddingRight(Integer value)
    {
        paddingRight = value;
    }

    public Integer paddingBottom()
    {
        return paddingBottom;
    }

    public void setPaddingBottom(Integer value)
    {
        paddingBottom = value;
    }

    public Integer paddingLeft()
    {
        return paddingLeft;
    }

    public void setPaddingLeft(Integer value)
    {
        paddingLeft = value;
    }

    public Integer margin()
    {
        return margin;
    }

    public void setMargin(Integer value)
    {
        margin = value;
    }

    public Integer marginTop()
    {
        return marginTop;
    }

    public void setMarginTop(Integer value)
    {
        marginTop = value;
    }

    public Integer marginRight()
    {
        return marginRight;
    }

    public void setMarginRight(Integer value)
    {
        marginRight = value;
    }

    public Integer marginBottom()
    {
        return marginBottom;
    }

    public void setMarginBottom(Integer value)
    {
        marginBottom = value;
    }

    public Integer marginLeft()
    {
        return marginLeft;
    }

    public void setMarginLeft(Integer value)
    {
        marginLeft = value;
    }

    public Integer gap()
    {
        return gap;
    }

    public void setGap(Integer value)
    {
        gap = value;
    }

    public String textAlign()
    {
        return textAlign;
    }

    public void setTextAlign(String value)
    {
        textAlign = value == null ? "" : value;
    }

    public String userSelect()
    {
        return userSelect;
    }

    public void setUserSelect(String value)
    {
        userSelect = value == null ? "" : value;
    }

    public Boolean scrollX()
    {
        return scrollX;
    }

    public void setScrollX(Boolean value)
    {
        scrollX = value;
    }

    public Boolean scrollY()
    {
        return scrollY;
    }

    public void setScrollY(Boolean value)
    {
        scrollY = value;
    }

    public TextGradient textColorGradient()
    {
        return textColorGradient;
    }

    public void setTextColorGradient(TextGradient value)
    {
        textColorGradient = value;
    }

    public TextGradient textBackgroundGradient()
    {
        return textBackgroundGradient;
    }

    public void setTextBackgroundGradient(TextGradient value)
    {
        textBackgroundGradient = value;
    }

    private static int parseInteger(String value)
    {
        return Integer.parseInt(value);
    }

    private static boolean parseBoolean(String value)
    {
        return "true".equalsIgnoreCase(value)
            || "1".equals(value)
            || "yes".equalsIgnoreCase(value)
            || "on".equalsIgnoreCase(value);
    }
}
